CRM/Core/Session.php

```diff
@@ -573,4 +585,60 @@ public function isEmpty() {
     return empty($_SESSION);
   }
 
+  /**
+   * Temporarily masquerade as a different user.
+   *
+   * Note: the override is stored in a static variable so does not persist past this page request.
+   *
+   * Returns a handle to identify this override so you can revert it later.
+   *
+   * @param int $contactId
+   *   Contact id to spoof as current user.
+   * @param string|int $handle
+   *   Specify your own handle if you wish.
+   * @return bool|string|int
+   *   Handle if anything was done, FALSE otherwise
+   */
+  public function overrideCurrentUser($contactId, $handle = NULL) {
+    if ($contactId == $this->get('userID')) {
+      // Current user is already $contactId; nothing to do
+      return FALSE;
+    }
+    $handle = $handle ?? uniqid();
+    self::$userOverride[$handle] = $contactId;
```

Comment:
1. Hmm, if there's going to be a static listing, then probably use `Civi::$statics`?

2. It seems like [this query](https://github.com/civicrm/org.civicrm.api4/pull/191/files#r320510727) ought to be invoked any time there's a transition of the active user ID, regardless of who/how/why it's transitioning.

3. Maybe this function should fire some kind of "set user" event? It feels a little brittle for any one party to enumerate all the things that happen when the user changes.

Reply:
TBH, part of me thinks that the whole of `CRM_Core_Config::singleton()` and `Civi::$statics` should be rebuilt/swapped whenever one switches the active user (to categorically prevent leaking data via caches). On the flip side, that's probably slower (esp if there are many user changes within the session), and those leak bugs are hypothetical. (*It's equally plausible from my POV that cache-leak vulnerabilities turn into long-term whackamole... or never happen at all...*) However, I'm very much on the fence about that - just want to verbalize to see what others think.

Reply:
The only one I've looked at is the contact ACL cache, which is on a per-user basis so it works fine with these overrides. But you're right there may be other spots in the system where things are cached assuming the logged in user id won't change.

My original use case for this is pretty limited though. I wanted an anonymous user, authenticated with a checksum, to be able to submit an afform as that contact. So the afform.submit call internally authenticates the checksum and then sets "acting_user" param on subsequent api calls. So that's not nearly as demanding a scenario as actually "masquerading" as a different user and clicking around the site.